src/share/util/shr_dmodel_mod.F90

```diff
@@ -872,6 +909,232 @@ subroutine shr_dmodel_readstrm(stream, pio_subsystem, pio_iotype, pio_iodesc, gs
+     write(logunit,F01) "shr_dmodel_readstrm_fullfile: not supported for iotype_std_netcdf"
+     call shr_sys_abort(subname//"ERROR extend shr_dmodel_readstrm_fullfile")
+
+  else
+
+     call t_startf(trim(lstr)//'_readpio')
+     call shr_mpi_bcast(sfldName,mpicom,'sfldName')
+     call shr_mpi_bcast(filename,mpicom,'filename')
+
+     call shr_stream_getCurrFile(stream,fileopen=fileopen,currfile=currfile,currpioid=pioid)
+
+     if (fileopen .and. currfile==filename) then
+        ! don't reopen file, all good
+     else
+        ! otherwise close the old file if open and open new file
+        if (fileopen) then
+           if (my_task == master_task) then
+              write(logunit,F00) 'close  : ',trim(currfile)
+              call shr_sys_flush(logunit)
+           endif
+           call pio_closefile(pioid)
+        endif
+        if (my_task == master_task) then
+           write(logunit,F00) 'open   : ',trim(filename)
+           call shr_sys_flush(logunit)
+        endif
+        rcode = pio_openfile(pio_subsystem, pioid, pio_iotype, trim(filename), pio_nowrite)
+        call shr_stream_setCurrFile(stream,fileopen=.true.,currfile=trim(filename),currpioid=pioid)
+
+        call pio_seterrorhandling(pioid,PIO_INTERNAL_ERROR)
+
+        rcode = pio_inq_varid(pioid,trim(sfldName),varid)
+        rcode = pio_inq_varndims(pioid, varid, ndims)
+        allocate(dimid(ndims))
+
+        rcode = pio_inq_vardimid(pioid, varid, dimid(1:ndims))
+
+        nx = 1
+        ny = 1
+        nz = 1
+        if (ndims >= 1) rcode = pio_inq_dimlen(pioid, dimid(1), nx)
+        if (ndims >= 2) rcode = pio_inq_dimlen(pioid, dimid(2), ny)
+        if (ndims >= 3) rcode = pio_inq_dimlen(pioid, dimid(3), nz)
+        deallocate(dimid)
+
+        if (gsize /= nx*ny .and. gsize /= nx*ny*nz) then
+           write(logunit,F01) "ERROR in data sizes ",nx,ny,nz,gsize
+           call shr_sys_abort(subname//"ERROR in data sizes")
+        endif
+
+        if (my_task == master_task) then
+           call shr_stream_getModelFieldList(stream,fldList)
+        endif
+        call shr_mpi_bcast(fldList,mpicom)
+
+        call mct_avect_clean(avFile)
+        call mct_aVect_init(avFile,rlist=fldList,lsize=nx*ny*nz)
+
+        call mct_gsmap_orderedPoints(gsMap,my_task,gsmOP)
+
+        allocate(count(3))
+        allocate(compDOF(lsize*nz),stat=rcode)
+        if (rcode /= 0) call shr_sys_abort(subname//"ERROR insufficient memory")
+
+        count(1) = nx
+        count(2) = ny
+        count(3) = nz
+
+        if (my_task == master_task) then
+           write(logunit,F02) 'file ' // trim(bstr) //': ',trim(filename),1,nz
+           call shr_sys_flush(logunit)
+        endif
+
+        cnt = 0
+        do n = 1,nz
+           do m = 1,lsize
+              cnt = cnt + 1
+              compDOF(cnt) = (n-1)*gsize + gsmOP(m)
+           enddo
+        enddo
+
+        call pio_initdecomp(pio_subsystem, pio_double, count, compDOF, pio_iodesc_local)
+
+        frame = 1
```

Comment:
`frame` is never incremented because all frames (`nx*ny*nz`)  defined in `count(1:3)` are read in a single call via `pio_read_darray()` at the same time a new netcdf file is opened in the `else` conditional ([shr_dmodel_mod.F90L1025](https://github.com/ESMCI/cime/pull/1418/files/8da53bdae38b55be332b08a16bdd13b6bce6d77c#diff-c3bed4dce38b944066024ec31df9aac2R1025))